lib/Monads/trace/Trace_RG.thy

```diff
@@ -471,38 +686,660 @@ lemma rely_cond_rtranclp:
+   \<Longrightarrow> \<lbrace>\<lambda>s0 s. (P \<longrightarrow> P' s0 s) \<and> P'' s0 s\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. P \<longrightarrow> Q rv s0 s\<rbrace>,\<lbrace>\<lambda>rv s0 s. P \<longrightarrow> E rv s0 s\<rbrace>"
+  unfolding validIE_def
+  by (rule rg_strengthen_post)
+     (fastforce intro!: rg_weak_lift_imp split: sum.splits)+
+
+lemma rg_weak_lift_impE_R:
+  "\<lbrakk>\<lbrace>P'\<rbrace>,\<lbrace>R\<rbrace> f -,\<lbrace>Q\<rbrace>,-; \<lbrace>P''\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<top>\<top>\<top>\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<lbrace>\<lambda>s0 s. (P \<longrightarrow> P' s0 s) \<and> P'' s0 s\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. P \<longrightarrow> Q rv s0 s\<rbrace>,-"
+  unfolding validIE_def
+  by (rule rg_strengthen_post)
+     (fastforce intro!: rg_weak_lift_imp split: sum.splits)+
+
+lemmas rg_weak_lift_imp_T = rg_weak_lift_imp[where G="\<top>\<top>" and P''="\<top>\<top>", simplified]
+lemmas rg_weak_lift_impE_T = rg_weak_lift_impE[where G="\<top>\<top>" and P''="\<top>\<top>", simplified]
+lemmas rg_weak_lift_impE_R_T = rg_weak_lift_impE_R[where G="\<top>\<top>" and P''="\<top>\<top>", simplified]
+
+lemma rg_vcg_ex_lift:
+  "\<lbrakk>\<And>x. \<lbrace>P x\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q x\<rbrace>\<rbrakk> \<Longrightarrow> \<lbrace>\<lambda>s0 s. \<exists>x. P x s0 s\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. \<exists>x. Q x rv s0 s\<rbrace>"
+  by (clarsimp simp: validI_def, blast)
+
+lemma rg_vcg_ex_liftE:
+  "\<lbrakk>\<And>x. \<lbrace>P x\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q x\<rbrace>,\<lbrace>E\<rbrace>\<rbrakk> \<Longrightarrow> \<lbrace>\<lambda>s0 s. \<exists>x. P x s0 s\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. \<exists>x. Q x rv s0 s\<rbrace>,\<lbrace>E\<rbrace>"
+  by (fastforce simp: validIE_def validI_def split: sum.splits)
+
+lemma rg_vcg_ex_liftE_E:
+  "\<lbrakk>\<And>x. \<lbrace>P x\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,-,\<lbrace>E x\<rbrace>\<rbrakk> \<Longrightarrow> \<lbrace>\<lambda>s0 s. \<exists>x. P x s0 s\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,-,\<lbrace>\<lambda>rv s0 s. \<exists>x. E x rv s0 s\<rbrace>"
+  by (fastforce simp: validIE_def validI_def split: sum.splits)
+
+lemma rg_vcg_ex_lift_R1:
+  "\<lbrakk>\<And>x. \<lbrace>P x\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,-\<rbrakk> \<Longrightarrow> \<lbrace>\<lambda>s0 s. \<exists>x. P x s0 s\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,-"
+  by (fastforce simp: validI_def validIE_def split: sum.splits)
+
+\<comment> \<open>FIXME MC: keep/name/move?\<close>
+lemma subst2:
+  "\<lbrakk>s = t; u = v; P s u\<rbrakk> \<Longrightarrow> P t v"
```

Comment:
Could go in `lib` -- it's for convenience, I assume.

Reply:
I used it to prove a goal of `P s0 (f s) ⟹ P (last_st_tr tr s0) (f s')`, where I was able to then show that each of the arguments were equal. I couldn't think of a better way to do this than writing this new version of `subst` and then doing `apply (erule subst2[rotated 2, where P=P])`.

If we're happy to keep it then it could go in `Monads/Monad_Lib.thy` I guess. `Monads` builds directly on `HOL`, so putting it in `lib` would lead to some weird dependencies.

Reply:
Yes, Monad_Lib is a good idea.